backend/pkg/executionlayer/reorg.go

```diff
@@ -0,0 +1,114 @@
+package executionlayer
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/gobitfly/beaconchain/pkg/commons/db2"
+	"github.com/gobitfly/beaconchain/pkg/commons/db2/database"
+	"github.com/gobitfly/beaconchain/pkg/commons/log"
+	"github.com/gobitfly/beaconchain/pkg/commons/types"
+)
+
+type EthClient interface {
+	HeaderByNumber(ctx context.Context, number *big.Int) (*gethtypes.Header, error)
+}
+
+type ReorgStore interface {
+	GetBlock(chainID string, number uint64) (*types.Eth1Block, error)
+	RevertBlock(chainID string, number uint64, blockHash []byte) error
+}
+
+type ReorgWatcher struct {
+	client         EthClient
+	store          ReorgStore
+	depth          uint64
+	chainID        string
+	lastBlockStore db2.LastBlocksStoreWriter
+}
+
+func NewReorgWatcher(client EthClient, store ReorgStore, depth uint64, chainID string, lastBlockStore db2.LastBlocksStoreWriter) *ReorgWatcher {
+	return &ReorgWatcher{
+		client:         client,
+		store:          store,
+		depth:          depth,
+		chainID:        chainID,
+		lastBlockStore: lastBlockStore,
+	}
+}
+
+func (r *ReorgWatcher) LookForReorg() error {
+	head, err := r.client.HeaderByNumber(context.Background(), nil)
+	if err != nil {
+		return err
+	}
+
+	// ensure we will not try to retrieve blocks that do not exist
+	depth := r.depth
+	if r.depth > head.Number.Uint64() {
+		depth = head.Number.Uint64()
+	}
+
+	ctx := context.Background()
+
+	// for each block check if block node hash and block db hash match
+	for i := head.Number.Uint64() - depth; i <= head.Number.Uint64(); i++ {
+		nodeBlock, err := r.client.HeaderByNumber(ctx, big.NewInt(int64(i)))
+		if err != nil {
+			return err
+		}
+		dbBlock, err := r.store.GetBlock(r.chainID, i)
+		if err != nil {
+			// exit if we hit a block that is not yet in the db
+			// it means that we never processed that block or that we revert that block
+			if errors.Is(err, database.ErrNotFound) {
+				return nil
```

Comment:
could be useful to add logging if block is not found in db
```suggestion
                                 log.Warnf("block %d not found in the database ", i)
				return nil
```

Reply:
No this the expected behavior, look at the comment
```
// exit if we hit a block that is not yet in the db
// it means that we never processed that block or that we revert that block
```
It would create noise for a normal process.